server/routerlicious/packages/services/src/mongoExceptionRetryRules/mongoError/index.ts

```diff
@@ -0,0 +1,172 @@
+        return error.code === 50
+            && error.errmsg
+            && (error.errmsg as string) === RequestTimedNoRateLimitInfo.errmsg;
+    }
+
+    shouldRetry: boolean = true;
+}
+
+// This handles request timeout from server with http info.
+class RequestTimedOutWithHttpInfo implements IMongoExceptionRetryRule {
+    private static readonly errmsgPrefix =
+        "Error=50, Details='Response status code does not indicate success: RequestTimeout (408);";
+    match(error: any): boolean {
+        return error.code === 50
+            && error.errmsg
+            && (error.errmsg as string).startsWith(RequestTimedOutWithHttpInfo.errmsgPrefix);
+    }
+
+    shouldRetry: boolean = true;
+}
+
+// This handles request timeout from server with additional rate limit info.
+class RequestTimedOutWithRateLimit implements IMongoExceptionRetryRule {
+    private static readonly codeName = "ExceededTimeLimit";
+    match(error: any): boolean {
+        // TODO: This timed out actually included two different messages:
+        // 1. Retries due to rate limiting: False.
+        // 2. Retries due to rate limiting: True.
+        // We might need to split this into two different rules after consult with DB team.
+        return error.code === 50
+            && error.errmsg
+            && (error.errmsg as string) === RequestTimedOutWithRateLimit.codeName;
+    }
+    shouldRetry: boolean = true;
+}
+
+// This handles server side temporary 503 issue
+class ServiceUnavailableRule implements IMongoExceptionRetryRule {
+    private static readonly errorDetails = "Response status code does not indicate success: ServiceUnavailable (503)";
+    match(error: any): boolean {
+        return error.code === 50
+            && error.errorDetails
+            && (error.errorDetails as string).includes(ServiceUnavailableRule.errorDetails);
+    }
+
+    shouldRetry: boolean = true;
+}
+
+// this handles the pool destroyed error from client side. Should relies on reconnect instead of retry?
+class TopologyDestroyed implements IMongoExceptionRetryRule {
+    private static readonly message = "Topology was destroyed";
+    match(error: any): boolean {
+        return error.code === 16
+            && error.message
+            && (error.message as string) === TopologyDestroyed.message;
+    }
+
+    shouldRetry: boolean = false;
+}
+
+// this handles the incorrect credentials set. Should not retry
+class UnUnauthorizedRule implements IMongoExceptionRetryRule {
+    private static readonly codeName = "Unauthorized";
+    match(error: any): boolean {
+        return error.code === 13
+            && error.codeName
+            && (error.codeName as string) === UnUnauthorizedRule.codeName;
+    }
+
+    shouldRetry: boolean = false;
+}
+
+// Maintain the list from more strick faster comparison to less strict slower comparison
+export const mongoErrorRetryRuleset: IMongoExceptionRetryRule[] = [
+    // The rules are using exactly equal
+    new InternalErrorRule(),
```

Comment:
Any reason why these classes need to be instantiated, instead of making all of them static? Apparently all classes above could be fully static but I might have missed something.